# Exploratory Data Analysis (EDA)

Prepared by Keina Aoki

November 2023

## Introduction

In today's digital age, social media and online news sources offers quick access to vast amounts of data. Yet the creation and dissemination of fake news poses an emerging challenge as it has the potential to generate mistrust in persons and institutions. There are broadly two types of fake information: misinformation and disinformation. 1) Misinformation refers to information which is incorrect or misleading or taken out of context unsuspectingly shared by people. 2) Disinformation refers to false information spread deliberately with the intent to deceive public opinion. While both mis/disinformation exist because people want to share a certain narrative that aligns with their interests and beliefs, disinformation is more dangerous and damaging because its intention is to destabilize societies. It is meant to influence people's beliefs, attitudes and consequentially, their behavior by purposely only providing one side of a narrative. Nonetheless, disinformation poses a threat to democracy because it reinforces polarizing viewpoints.

The authenticity of news content has become difficult to discern not only because of social media algorithms that determine the content a user sees and interacts with, but because human minds are faulty to judgement about information. Thus, it is imperative to develop tools that can detect fake news to ensure that we are well informed.

This project aims to train a model to identify fake news through twitter posts (now known as X). It will employ machine learning techniques such as natural language processing and eventually use sentiment analysis to assess the credibility of a news story based on the emotion that the words in the statement elicits. The data set is sourced from https://www.unb.ca/cic/datasets/truthseeker-2023.html which includes fake and real news.

In this notebook, we explore and perform Exploratory Data Analysis (EDA) to uncover valuable insights from the data such as patterns and trends.

## Data Dictionary

Our dataset (Features_For_Traditional_ML.csv) includes the following fields:


| Name | Description   |
|:------|:------|
|  unique_count |number of unique, complex words|
|total_count |total number of words|
|ORG_percent | Percent of text including spaCy ORG tags (organizations, companies, agencies, institutions)
|NORP_percent |Percent of text including spaCy NORP tags (nationalities or religious or political groups)
|GPE_percent | Percent of text including spaCy GPE tags (countries, cities, states)
|PERSON_percent | Percent of text including spaCy PERSON tags (people, including fictional)
|MONEY_percent | Percent of text including spaCy MONEY tags (monetary values, including unit)
|DATE_percent | Percent of text including spaCy DATE tags (absolute or relative dates or periods)
|CARDINAL_percent | Percent of text including spaCy CARDINAL tags (numerals that do not fall under another type)
|PERCENT_percent | Percent of text including spaCy PERCENT tags (percentage, including "%")
|ORDINAL_percent | Percent of text including spaCy ORDINAL tags ("first", "second", etc.)
|FAC_percent | Percent of text including spaCy FAC tags (buildings, airports, highways, bridges, etc.)
|LAW_percent | Percent of text including spaCy LAW tags (named documents made into laws)
|PRODUCT_percent | Percent of text including spaCy PRODUCT tags (objects, vehicles, foods, etc. (not services))
|EVENT_percent | Percent of text including spaCy EVENT tags (named hurricanes, battles, wars, sports events, etc.)
|TIME_percent | Percent of text including spaCy TIME tags (times smaller than a day)
|LOC_percent | Percent of text including spaCy LOC tags (non-GPE locations, mountain ranges, bodies of water)
|WORK_OF_ART_percent | Percent of text including spaCy WOA tags (titles of books, songs, etc.)
|QUANTITY_percent | Percent of text including spaCy QUANTITY tags (measurements, as of weight or distance) 
|LANGUAGE_percent | Percent of text including spaCy LANGUAGE tags (any named language)
|Max Word | Length of the longest word in the sentence
|Min Word | Length of the shortest word in the sentence
|Avg Word Length | Average length of words in the sentence
|present_verb | Number of present tense verbs
|past_verb | Number of past tense verbs
|adjectives | Number of adjectives
|adverbs | Number of adverbs
|adpositions | Number of adpositions
|pronouns | Number of pronouns
|TO’s | Number of to usages
|determiners | Number of determiners
|conjunctions | Number of conjunctions
|dots | Number of (.) used
|exclamations | Number of (!) used
|question | Number of (?) used
|ampersand | Number of (&) used
|capitals | Number of capitalized letters
|digits | Number of digits (0-9) used
|word_count | Number of words
|long_word_freq | Number of long words
|short_word_freq | Number of short words
|followers_count | Number of followers
|friends_count | Number of friends
|favourites_count | Number of favourites across all tweets
|statuses_count | Number of tweets
|listed_count | Number of tweets the user has in lists
|mentions | Number of times the user was mentioned
|replies | Number of replies the user has
|retweets | Number of retweets the user has
|favourites | Number of favourites the user has
|hashtags | Number of hashtags (#) the user has used
|URLs | whether the user has a provided a url in relation to their profile
|quotes | Number of times the user has been quote tweeted
|BotScoreBinary | Binary score whether the user is considered a bot or not
|cred | Credibility score
|normalized_influence | Influence score the user has, normalized
|majority_target | Truth value of the tweet
|statement | Headline of a new article
|BinaryNumTarget | Binary representation of the statement's truth value (1 = True / 0 = False)
|tweet | Twitter posts related to the associated manual keywords

Note: spaCy tags refers to the spaCy natural language processing model that classifies text into categories

## Part 1: Data Exploration

In [34]:
# importing required libraries:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

In [35]:
# read data from csv file:
df = pd.read_csv("Features_For_Traditional_ML_Techniques.csv")

In [36]:
# sanity checks with the columns and the first 5 rows
pd.set_option('display.max_columns', None)
df.head()

Unnamed: 0  majority_target  \
0           0             True   
1           1             True   
2           2             True   
3           3             True   
4           4             True   

                                           statement  BinaryNumTarget  \
0  End of eviction moratorium means millions of A...              1.0   
1  End of eviction moratorium means millions of A...              1.0   
2  End of eviction moratorium means millions of A...              1.0   
3  End of eviction moratorium means millions of A...              1.0   
4  End of eviction moratorium means millions of A...              1.0   

                                               tweet  followers_count  \
0  @POTUS Biden Blunders - 6 Month Update\n\nInfl...           4262.0   
1  @S0SickRick @Stairmaster_ @6d6f636869 Not as m...           1393.0   
2  THE SUPREME COURT is siding with super rich pr...              9.0   
3  @POTUS Biden Blunders\n\nBroken campaign promi...           4262.0   
4  @OhComfy I agree. The confluence of events rig...             70.0   

   friends_count  favourites_count  statuses_count  listed_count  following  \
0         3619.0           34945.0         16423.0          44.0        0.0   
1         1621.0           31436.0         37184.0          64.0        0.0   
2           84.0             219.0          1184.0           0.0        0.0   
3         3619.0           34945.0         16423.0          44.0        0.0   
4          166.0           15282.0          2194.0           0.0        0.0   

                              embeddings  BotScore  BotScoreBinary      cred  \
0  [[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]]      0.03             0.0  0.540794   
1  [[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]]      0.03             0.0  0.462177   
2  [[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]]      0.03             0.0  0.096774   
3  [[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]]      0.03             0.0  0.540794   
4  [[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]]      0.03             0.0  0.296610   

   normalize_influence  mentions  quotes  replies  retweets  favourites  \
0             0.104602       1.0     1.0      1.0       3.0        10.0   
1             0.094436       3.0     0.0      0.0       0.0         1.0   
2             0.039846       0.0     2.0      5.0       0.0         0.0   
3             0.104602       1.0     0.0      0.0       0.0         0.0   
4             0.061135       1.0     0.0      0.0       0.0         1.0   

   hashtags  URLs  unique_count  total_count  ORG_percentage  NORP_percentage  \
0       0.0   1.0             5            6        0.333333         0.166667   
1       0.0   1.0             2            4        0.000000         0.250000   
2       0.0   1.0             1            1        0.000000         1.000000   
3       0.0   1.0             4            6        0.000000         0.333333   
4       0.0   1.0             3            3        0.333333         0.000000   

   GPE_percentage  PERSON_percentage  MONEY_percentage  DATE_percentage  \
0        0.166667           0.166667          0.166667         0.000000   
1        0.000000           0.750000          0.000000         0.000000   
2        0.000000           0.000000          0.000000         0.000000   
3        0.000000           0.166667          0.166667         0.333333   
4        0.333333           0.000000          0.333333         0.000000   

   CARDINAL_percentage  PERCENT_percentage  ORDINAL_percentage  \
0                  0.0                 0.0                 0.0   
1                  0.0                 0.0                 0.0   
2                  0.0                 0.0                 0.0   
3                  0.0                 0.0                 0.0   
4                  0.0                 0.0                 0.0   

   FAC_percentage  LAW_percentage  PRODUCT_percentage  EVENT_percentage  \
0             0.0             0.0                 0.0               0.0   
1             0.0           

In [37]:
# remove the column 'Unnamed: 0' as it is just an index
df = df.drop(['Unnamed: 0'], axis=1)

# remove the collowing as there are no documentation on it and may skew the data later on
df = df.drop(['following'], axis=1)
df = df.drop(['embeddings'], axis=1)
df = df.drop(['BotScore'], axis=1)

In [38]:
# sanity check after dropping columns
df.head()

majority_target                                          statement  \
0             True  End of eviction moratorium means millions of A...   
1             True  End of eviction moratorium means millions of A...   
2             True  End of eviction moratorium means millions of A...   
3             True  End of eviction moratorium means millions of A...   
4             True  End of eviction moratorium means millions of A...   

   BinaryNumTarget                                              tweet  \
0              1.0  @POTUS Biden Blunders - 6 Month Update\n\nInfl...   
1              1.0  @S0SickRick @Stairmaster_ @6d6f636869 Not as m...   
2              1.0  THE SUPREME COURT is siding with super rich pr...   
3              1.0  @POTUS Biden Blunders\n\nBroken campaign promi...   
4              1.0  @OhComfy I agree. The confluence of events rig...   

   followers_count  friends_count  favourites_count  statuses_count  \
0           4262.0         3619.0           34945.0         16423.0   
1           1393.0         1621.0           31436.0         37184.0   
2              9.0           84.0             219.0          1184.0   
3           4262.0         3619.0           34945.0         16423.0   
4             70.0          166.0           15282.0          2194.0   

   listed_count  BotScoreBinary      cred  normalize_influence  mentions  \
0          44.0             0.0  0.540794             0.104602       1.0   
1          64.0             0.0  0.462177             0.094436       3.0   
2           0.0             0.0  0.096774             0.039846       0.0   
3          44.0             0.0  0.540794             0.104602       1.0   
4           0.0             0.0  0.296610             0.061135       1.0   

   quotes  replies  retweets  favourites  hashtags  URLs  unique_count  \
0     1.0      1.0       3.0        10.0       0.0   1.0             5   
1     0.0      0.0       0.0         1.0       0.0   1.0             2   
2     2.0      5.0       0.0         0.0       0.0   1.0             1   
3     0.0      0.0       0.0         0.0       0.0   1.0             4   
4     0.0      0.0       0.0         1.0       0.0   1.0             3   

   total_count  ORG_percentage  NORP_percentage  GPE_percentage  \
0            6        0.333333         0.166667        0.166667   
1            4        0.000000         0.250000        0.000000   
2            1        0.000000         1.000000        0.000000   
3            6        0.000000         0.333333        0.000000   
4            3        0.333333         0.000000        0.333333   

   PERSON_percentage  MONEY_percentage  DATE_percentage  CARDINAL_percentage  \
0           0.166667          0.166667         0.000000                  0.0   
1           0.750000          0.000000         0.000000                  0.0   
2           0.000000          0.000000         0.000000                  0.0   
3           0.166667          0.166667         0.333333                  0.0   
4           0.000000          0.333333         0.000000                  0.0   

   PERCENT_percentage  ORDINAL_percentage  FAC_percentage  LAW_percentage  \
0                 0.0                 0.0             0.0             0.0   
1                 0.0                 0.0             0.0             0.0   
2                 0.0                 0.0             0.0             0.0   
3                 0.0                 0.0             0.0             0.0   
4                 0.0                 0.0             0.0             0.0   

   PRODUCT_percentage  EVENT_percentage  TIME_percentage  LOC_percentage  \
0                 0.0               0.0              0.0             0.0   
1                 0.0               0.0              0.0             0.0   
2                 0.0               0.0              0.0             0.0   
3                 0.0               0.0              0.0             0.0   
4                 0.0               0.0              0.0             

In [39]:
# sanity checks with the last 5 rows
df.tail()

majority_target                                          statement  \
134193            False  Joe Bidens great-grandfather Joseph J. Biden w...   
134194            False  Joe Bidens great-grandfather Joseph J. Biden w...   
134195            False  Joe Bidens great-grandfather Joseph J. Biden w...   
134196            False  Joe Bidens great-grandfather Joseph J. Biden w...   
134197            False  Joe Bidens great-grandfather Joseph J. Biden w...   

        BinaryNumTarget                                              tweet  \
134193              0.0  Joe Biden's family owned African slaves....\n\...   
134194              0.0  Joe Bidens great, great grandfather was a slav...   
134195              0.0  @ChevyChaseToGo "Joe Bidens great-grandfather ...   
134196              0.0  @JoeBiden Facts are Bidens VP Kamala Harris Gr...   
134197              0.0  @sunny Yes representation matters. Did you kno...   

        followers_count  friends_count  favourites_count  statuses_count  \
134193           2075.0         2662.0          156306.0         62375.0   
134194           1636.0         1626.0          215290.0         44011.0   
134195            286.0         2501.0           28377.0          7617.0   
134196              2.0           45.0              16.0           509.0   
134197            178.0          471.0            1254.0          2094.0   

        listed_count  BotScoreBinary      cred  normalize_influence  mentions  \
134193           3.0             0.0  0.438041             0.098915       0.0   
134194          10.0             0.0  0.501533             0.096203       0.0   
134195           1.0             0.0  0.102619             0.076870       1.0   
134196           0.0             0.0  0.042553             0.024038       1.0   
134197           0.0             0.0  0.274268             0.070342       1.0   

        quotes  replies  retweets  favourites  hashtags  URLs  unique_count  \
134193     1.0      1.0       1.0         2.0       0.0   1.0             1   
134194     1.0      0.0       1.0         1.0       0.0   0.0             2   
134195     0.0      0.0       0.0         0.0       0.0   1.0             2   
134196     0.0      0.0       0.0         0.0       0.0   1.0             0   
134197     0.0      0.0       0.0         0.0       0.0   1.0             2   

        total_count  ORG_percentage  NORP_percentage  GPE_percentage  \
134193            3        1.000000              0.0        0.000000   
134194            2        0.500000              0.0        0.000000   
134195            2        0.000000              0.0        0.000000   
134196            0        0.000000              0.0        0.000000   
134197            3        0.333333              0.0        0.666667   

        PERSON_percentage  MONEY_percentage  DATE_percentage  \
134193                0.0               0.0              0.0   
134194                0.5               0.0              0.0   
134195                0.0               0.5              0.5   
134196                0.0               0.0              0.0   
134197                0.0               0.0              0.0   

        CARDINAL_percentage  PERCENT_percentage  ORDINAL_percentage  \
134193                  0.0                 0.0                 0.0   
134194                  0.0                 0.0                 0.0   
134195                  0.0                 0.0                 0.0   
134196                  0.0                 0.0                 0.0   
134197                  0.0                 0.0                 0.0   

        FAC_percentage  LAW_percentage  PRODUCT_percentage  EVENT_percentage  \
134193             0.0             0.0                 0.0               0.0   
134194             0.0             0.0                 0.0               0.0   
134195             0.0             0.0                 0.0               0.0   
134196             0.0             0.0                 0.0               0.0   
134197

In [40]:
# display dimensions of data:
df.shape
print(f'data includes {df.shape[0]} rows and {df.shape[1]} columns.')

data includes 134198 rows and 60 columns.


In [41]:
# summary of Features_For_Traditional_ML_Techniques:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 134198 entries, 0 to 134197
Data columns (total 60 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   majority_target         134198 non-null  bool   
 1   statement               134198 non-null  object 
 2   BinaryNumTarget         134198 non-null  float64
 3   tweet                   134198 non-null  object 
 4   followers_count         134198 non-null  float64
 5   friends_count           134198 non-null  float64
 6   favourites_count        134198 non-null  float64
 7   statuses_count          134198 non-null  float64
 8   listed_count            134198 non-null  float64
 9   BotScoreBinary          134198 non-null  float64
 10  cred                    134198 non-null  float64
 11  normalize_influence     134198 non-null  float64
 12  mentions                134198 non-null  float64
 13  quotes                  134198 non-null  float64
 14  replies             

Findings:
- All the columns have 134198.
- df contains 57 numerical columns. 22 of the columns are integer, the rest are float. 
- df contains 3 categorical columns. This includes the statements and twees which are objects. Majority target is a boolean.

In [42]:
df.describe()

BinaryNumTarget  followers_count  friends_count  favourites_count  \
count    134198.000000     1.341980e+05  134198.000000      1.341980e+05   
mean          0.513644     1.129308e+04    1893.454455      3.298123e+04   
std           0.499816     4.374971e+05    6997.695671      6.878021e+04   
min           0.000000     0.000000e+00       0.000000      0.000000e+00   
25%           0.000000     7.000000e+01     168.000000      1.356000e+03   
50%           1.000000     3.540000e+02     567.000000      8.377000e+03   
75%           1.000000     1.573000e+03    1726.000000      3.352650e+04   
max           1.000000     1.306019e+08  586901.000000      1.765080e+06   

       statuses_count   listed_count  BotScoreBinary           cred  \
count    1.341980e+05  134198.000000   134198.000000  134198.000000   
mean     3.419576e+04      73.300198        0.032355       0.405852   
std      7.510120e+04    1083.274277        0.176942       0.239223   
min      1.000000e+00       0.000000        0.000000       0.000000   
25%      3.046000e+03       0.000000        0.000000       0.223684   
50%      1.101900e+04       2.000000        0.000000       0.386691   
75%      3.357375e+04      11.000000        0.000000       0.525020   
max      2.958918e+06  222193.000000        1.000000       1.000000   

       normalize_influence       mentions         quotes        replies  \
count        134198.000000  134198.000000  134198.000000  134198.000000   
mean              0.077665       1.388918       0.573406       1.914201   
std               0.026184       1.471604      28.436726     122.041183   
min               0.000000       0.000000       0.000000       0.000000   
25%               0.061814       0.000000       0.000000       0.000000   
50%               0.079436       1.000000       0.000000       0.000000   
75%               0.095308       2.000000       0.000000       1.000000   
max               0.208606      12.000000    5913.000000   42068.000000   

            retweets     favourites       hashtags           URLs  \
count  134198.000000  134198.000000  134198.000000  134198.000000   
mean        6.674354      27.572386       0.104726       0.737701   
std       406.542579    1831.425703       0.458687       0.439886   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       1.000000   
75%         0.000000       1.000000       0.000000       1.000000   
max    126062.000000  460320.000000      12.000000       1.000000   

        unique_count    total_count  ORG_percentage  NORP_percentage  \
count  134198.000000  134198.000000   134198.000000    134198.000000   
mean        2.365624       3.441229        0.199964         0.067412   
std         1.351617       2.479373        0.276376         0.165502   
min         0.000000       0.000000        0.000000         0.000000   
25%         1.000000       2.000000        0.000000         0.000000   
50%         2.000000       3.000000        0.000000         0.000000   
75%         3.000000       5.000000        0.333333         0.000000   
max         9.000000      32.000000        1.000000         1.000000   

       GPE_percentage  PERSON_percentage  MONEY_percentage  DATE_percentage  \
count   134198.000000      134198.000000     134198.000000    134198.000000   
mean         0.136151           0.221930          0.029525         0.102339   
std          0.242561           0.298133          0.106880         0.206059   
min          0.000000           0.000000          0.000000         0.000000   
25%          0.000000           0.000000          0.000000         0.000000   
50%          0.000000           0.000000          0.000000         0.000000   
75%          0.222222           0.333333          0.000000         0.142857   
max          1.000000           1.000000          1.000000         1.000000   

       CARDINAL_percent

In [43]:
# check the number of missing values for columns:
df.isna().sum()

majority_target           0
statement                 0
BinaryNumTarget           0
tweet                     0
followers_count           0
friends_count             0
favourites_count          0
statuses_count            0
listed_count              0
BotScoreBinary            0
cred                      0
normalize_influence       0
mentions                  0
quotes                    0
replies                   0
retweets                  0
favourites                0
hashtags                  0
URLs                      0
unique_count              0
total_count               0
ORG_percentage            0
NORP_percentage           0
GPE_percentage            0
PERSON_percentage         0
MONEY_percentage          0
DATE_percentage           0
CARDINAL_percentage       0
PERCENT_percentage        0
ORDINAL_percentage        0
FAC_percentage            0
LAW_percentage            0
PRODUCT_percentage        0
EVENT_percentage          0
TIME_percentage           0
LOC_percentage      

There are no missing values.

In [44]:
# check for exact duplicates in rows:
df.duplicated().sum()

0

There are no duplicate rows.

## Part 2: EDA

Here we will being exploring the data to gain insights and identify patterns and trends:

### Correlation

In [45]:
# check correlation of numerical data
df_num = df.select_dtypes(include = "number")
df_num.corr()

BinaryNumTarget  followers_count  friends_count  \
BinaryNumTarget                1.000000         0.013712       0.029014   
followers_count                0.013712         1.000000       0.206958   
friends_count                  0.029014         0.206958       1.000000   
favourites_count               0.044088        -0.001963       0.174073   
statuses_count                 0.077732         0.031071       0.242490   
listed_count                   0.033798         0.815270       0.183237   
BotScoreBinary                -0.012912        -0.004687      -0.046619   
cred                           0.111584         0.055394       0.085949   
normalize_influence            0.131866         0.083034       0.330636   
mentions                      -0.049086        -0.017852      -0.048792   
quotes                        -0.007470         0.079789       0.026679   
replies                       -0.005278         0.054423       0.015716   
retweets                      -0.002832         0.071994       0.019105   
favourites                    -0.003763         0.071982       0.016180   
hashtags                       0.105710         0.007529       0.032510   
URLs                          -0.136375        -0.009708      -0.028249   
unique_count                   0.071814         0.001381       0.004991   
total_count                    0.040115         0.004600       0.002637   
ORG_percentage                -0.038610         0.000240       0.003323   
NORP_percentage                0.057028        -0.001922      -0.003502   
GPE_percentage                 0.047484        -0.001564       0.000792   
PERSON_percentage             -0.111567        -0.005296       0.006480   
MONEY_percentage               0.077220         0.000207      -0.004051   
DATE_percentage                0.025115        -0.000400      -0.004151   
CARDINAL_percentage            0.075818         0.010215       0.004322   
PERCENT_percentage             0.038205         0.004105      -0.004220   
ORDINAL_percentage             0.010510         0.000424      -0.001920   
FAC_percentage                -0.022373        -0.001210      -0.002813   
LAW_percentage                 0.029426        -0.000223       0.001866   
PRODUCT_percentage            -0.006000        -0.000481       0.000963   
EVENT_percentage               0.021909         0.001270       0.000316   
TIME_percentage               -0.020769        -0.001262       0.001726   
LOC_percentage                 0.007956        -0.002079      -0.003306   
WORK_OF_ART_percentage        -0.020495        -0.001063      -0.000950   
QUANTITY_percentage           -0.009830        -0.000013      -0.000681   
LANGUAGE_percentage            0.038365        -0.000594       0.002948   
Word count                    -0.079875         0.002037      -0.004035   
Max word length               -0.012994        -0.000339      -0.005967   
Min word length                0.011125         0.002193       0.000259   
Average word length            0.045793         0.003860       0.004759   
present_verbs                 -0.092082         0.000161      -0.003704   
past_verbs                    -0.092284         0.000456      -0.002083   
adjectives                     0.052809         0.001941      -0.003082   
adverbs                       -0.049680         0.001267      -0.001503   
adpositions                    0.022727        -0.000124      -0.003366   
pronouns                      -0.114162        -0.003215      -0.004953   
TOs                           -0.051943         0.000916      -0.003690   
determiners                   -0.024640         0.003797      -0.001293   
conjunctions                  -0.045761        -0.002829      -0.003196   
dots                          -0.024325        -0.002065      -0.001201   
exclamation                   -0.068000        -0.002482      -0.001837   
questions                     -0.044242        -0.001470      -0.002834   
ampersand                     -0.004879        -0

In [ ]:
# display the heatmap for correlation matrix:
plt.figure(figsize= (30, 32))
sns.heatmap(round(df_num.corr(),2), cmap='coolwarm', annot=True)

<Axes: >

Findings:
- normalize_influence (influence score that the user has) and cred (credibility score) has a positive correlation of approximately 0.63. This is expected as users with high credibility are more likely to have a higher influence score.
- listed count (number of tweets the user has in lists) and followers count has a strong positive correlation of approximately 0.82
- quotes have a positive correlation with replies (0.74), retweets (0.78), and favourites (0.76). This could be explained by the fact that quotes refer to the "Number of times the user has been quote tweeted" which could be in the from of replies to that tweet, retweeting that exact tweet, or favourited.
- There is also a very high positive correlation between retweets and favourites (0.95), as these two have similar functions in liking or sharing a particular tweet.
- the total count (total number of words) and unique count (total number of unique words) has a positive correlation of approximately 0.81. 
- word count and short word frequency has a positive correlation of approximately 0.83. This means that the word count of a tweet are composed of seveal short words.

### Create Scatterplots for Columns with High Correlation

In [ ]:
# scatterplot for influence score and credibility score
plt.figure()
sns.scatterplot(x="normalize_influence", y="cred", data=df)
plt.title("Influence Score by Credibility")
plt.xlabel("Influence Score Normalized")
plt.ylabel("Credibility Score")
plt.show()

Finding:
- The highest credibility score that can be assigned to a user is 1.
- There appears to be a positive correlation where an increase in either the credibility score or influence score increases the other variable.

In [ ]:
# scatterplot for listed count and follower count
plt.figure()
sns.scatterplot(x="listed_count", y="followers_count", hue="BinaryNumTarget",data=df)
plt.xlabel("Number of Tweets User has in List")
plt.ylabel("Number of Followers")
plt.legend(labels = ["Fake", "Real"])
plt.show()

Finding:
- The scatterplot displays a slight positive correlation, however it is mostly characterized by the outliers.
- Fake news has more outliers compared to Real news.

In [ ]:
# scatterplot for quotes and replies
plt.figure()
sns.scatterplot(x="quotes", y="replies", hue="BinaryNumTarget",data=df)
plt.xlabel("Number of Times User has been Quote Tweeted")
plt.ylabel("Number of Replies the User Has")
plt.legend(labels = ["Fake", "Real"])
plt.show()

Finding:
- The scatterplot displays a slight positive correlation.
- The datapoints that make up the linear correlation are mostly Real news indicating that there is a positive relationship between the number of replies a user has the number of times a user has been quoted.

In [ ]:
# scatterplot for quotes and retweets
plt.figure()
sns.scatterplot(x="quotes", y="retweets", hue="BinaryNumTarget",data=df)
plt.xlabel("Number of Times User has been Quote Tweeted")
plt.ylabel("Number of Retweets the User Has")
plt.legend(labels = ["Fake", "Real"])
plt.show()

Finding:
- The scatterplot displays a slight positive correlation.
- The datapoints that make up the linear correlation are mostly Real news indicating that there is a positive relationship between the number of retweets a user has the number of times a user has been quoted.

In [ ]:
# scatterplot for quotes and favourites
plt.figure()
sns.scatterplot(x="quotes", y="favourites", hue="BinaryNumTarget",data=df)
plt.xlabel("Number of Times User has been Quote Tweeted")
plt.ylabel("Number of Favourites the User Has")
plt.legend(labels = ["Fake", "Real"])
plt.show()

Finding:
- The scatterplot displays a slight positive correlation.
- The datapoints that make up the linear correlation are mostly Real news indicating that there is a positive relationship between the number of favourites a user has the number of times a user has been quoted.

In [ ]:
# scatterplot for retweet and favourites
plt.figure()
sns.scatterplot(x="retweets", y="favourites", hue="BinaryNumTarget",data=df)
plt.xlabel("Number of Retweets the User HAs")
plt.ylabel("Number of Favourties the User Has")
plt.legend(labels = ["Fake", "Real"])
plt.show()

Finding:
- The scatterplot displays a slight positive correlation. However, it is very difficult to tell given that most of the datapoints are clustered.
- The outliers are datapoints from Real News.

In [ ]:
# scatterplot for total count and unique count
plt.figure()
sns.scatterplot(x="total_count", y="unique_count", hue="BinaryNumTarget",data=df)
plt.xlabel("Total Number of Words")
plt.ylabel("Total Number of Unique Words")
plt.legend(labels = ["Fake", "Real"])
plt.show()

Finding:
- The scatterplot does not display a linear correlation between the total number of unique words and total number of words.

In [ ]:
# scatterplot for word count and short words
plt.figure()
sns.scatterplot(x="Word count", y="short_word_freq", hue="BinaryNumTarget",data=df)
plt.xlabel("Number Words")
plt.ylabel("Frequency of Short Words")
plt.legend(labels = ["Fake", "Real"])
plt.show()

Findings:
- The number of words does not exhibit a notable linear correlation with the frequency of short words.

### EDA on Fake vs. Real News

In [ ]:
# check the total number of fake and real news
df["BinaryNumTarget"].value_counts()

In [ ]:
plt.figure()
tgtcnt = sns.countplot(x="BinaryNumTarget", data=df, color="coral")
tgtcnt.set(title="Number of Fake vs. Real News")
tgtcnt.set_xticks([0,1])
tgtcnt.set_xticklabels(["Fake", "Real"])
plt.xlabel("Target")
plt.show()

Finding:
- There are more real news than fake news in this dataset.

In [ ]:
# plot the word count in statement by target
plt.figure()
wrdcnt = sns.boxplot(x="BinaryNumTarget", y="Word count", data=df, color="coral")
wrdcnt.set(title = "Number of words by Target")
wrdcnt.set_xticks([0,1])
wrdcnt.set_xticklabels(["Fake", "Real"])
plt.xlabel("Target")
plt.show()

Finding:
- The median for the number of words in the news does not differ very much between Fake and Real News.

In [ ]:
# plot the number of short words by target
plt.figure()
shrt_freq = sns.barplot(y="short_word_freq", x="BinaryNumTarget", data=df, color="coral", errorbar=None)
shrt_freq.set(title= "Number of Short Words in News by Target")
shrt_freq.set_ylabel("Number of Short Words")
shrt_freq.set_xticks([0,1])
shrt_freq.set_xticklabels(["Fake", "Real"])
plt.xlabel("Target")
plt.show()

Finding:
- There are more short words in fake news compared to real news.

In [ ]:
# plot the number of long words by target
plt.figure()
shrt_freq = sns.barplot(y="long_word_freq", x="BinaryNumTarget", data=df, color="coral", errorbar=None)
shrt_freq.set(title= "Number of Long Words in News by Target")
shrt_freq.set_ylabel("Number of Long Words")
shrt_freq.set_xticks([0,1])
shrt_freq.set_xticklabels(["Fake", "Real"])
plt.xlabel("Target")
plt.show()

Finding:
- There are more long words in fake news compared to real news.

In [ ]:
# split data by target
fake = df[(df["BinaryNumTarget"] == 0)]
real = df[(df["BinaryNumTarget"] == 1)]

In [ ]:
# distribution of word count in Fake News
plt.figure()
plt.hist(fake["Word count"])
plt.axvline(fake["Word count"].mean(), c='black', label='Mean')
plt.title("Distribution of Word Count on Fake News")
plt.xlabel("Word count")
plt.ylabel("Frequency")
plt.show()

In [ ]:
# distribution of word count in Real News
plt.figure()
plt.hist(real["Word count"])
plt.axvline(real["Word count"].mean(), c='black', label='Mean')
plt.title("Distribution of Word Count on Real News")
plt.xlabel("Word count")
plt.ylabel("Frequency")
plt.show()

Finding:
- The frequency of the word count in fake news is more normalized compared to the real news.
- The fake news shows a greater left skew.

In [ ]:
# plot the number of followers by target
plt.figure()
shrt_freq = sns.barplot(y="followers_count", x="BinaryNumTarget", data=df, color="coral", errorbar=None)
shrt_freq.set(title= "Number of Followers by Target")
shrt_freq.set_ylabel("Number of Followers")
shrt_freq.set_xticks([0,1])
shrt_freq.set_xticklabels(["Fake", "Real"])
plt.xlabel("Target")
plt.show()

Finding:
- Users posting real news has more followers comapred to users posting fake news.

In [ ]:
# plot the influence score by target
plt.figure()
shrt_freq = sns.barplot(y="normalize_influence", x="BinaryNumTarget", data=df, color="coral", errorbar=None)
shrt_freq.set(title= "Total Influence Scores by Target")
shrt_freq.set_ylabel("Influence Scores")
shrt_freq.set_xticks([0,1])
shrt_freq.set_xticklabels(["Fake", "Real"])
plt.xlabel("Target")
plt.show()

Finding:
- Users posting real news has a greater influence score.# plot the number of followers by target.

In [ ]:
# plot the number of retweets by target
plt.figure()
shrt_freq = sns.barplot(y="retweets", x="BinaryNumTarget", data=df, color="coral", errorbar=None)
shrt_freq.set(title= "Number of Retweets by Target")
shrt_freq.set_ylabel("Number of Retweets")
shrt_freq.set_xticks([0,1])
shrt_freq.set_xticklabels(["Fake", "Real"])
plt.xlabel("Target")
plt.show()

Finding:
- Fake news has more retweets compared to real news.

In [ ]:
# plot the number of replies by target
plt.figure()
shrt_freq = sns.barplot(y="replies", x="BinaryNumTarget", data=df, color="coral", errorbar=None)
shrt_freq.set(title= "Number of Replies by Target")
shrt_freq.set_ylabel("Number of Replies")
shrt_freq.set_xticks([0,1])
shrt_freq.set_xticklabels(["Fake", "Real"])
plt.xlabel("Target")
plt.show()

Finding:
- Fake news has a greater number of replies comapred to real news.